diff -ru mitsuba.original/src/films/mfilm.cpp mitsuba.patched/src/films/mfilm.cpp
--- mitsuba.original/src/films/mfilm.cpp	2014-07-01 23:45:54.000000000 +0200
+++ mitsuba.patched/src/films/mfilm.cpp	2014-07-15 14:27:05.759275966 +0200
@@ -298,10 +298,12 @@
 						if (m_fileFormat == EMATLAB) {
 							os << *ptr;
 						} else {
+							/* Mathematica uses the peculiar '*^' notation rather than the standard 'e' notation. */
 							std::ostringstream oss;
+							oss << std::setprecision(m_digits);
 							oss << *ptr;
 							std::string str = oss.str();
-							boost::replace_first(str, "e", " * 10^");
+							boost::replace_first(str, "e", "*^");
 							os << str;
 						}
 
